src/ptyfwd/ptyfwd.c

```diff
@@ -0,0 +1,190 @@
+        assert(f);
+
+        if (rcode == -ECANCELED) {
+                log_debug_errno(rcode, "PTY forwarder disconnected.");
+                return sd_event_exit(e, EXIT_SUCCESS);
+        } else if (rcode < 0) {
+                (void) sd_event_exit(e, EXIT_FAILURE);
+                return log_error_errno(rcode, "Error on PTY forwarding logic: %m");
+        }
+
+        return 0;
+}
+
+static int helper_on_exit(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        /* Add 128 to signal exit statuses to mimick shells. */
+        return sd_event_exit(sd_event_source_get_event(s), si->si_status + (si->si_code == CLD_EXITED ? 0 : 128));
+}
+
+static int run(int argc, char *argv[]) {
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        _cleanup_close_ int pty_fd = -EBADF, peer_fd = -EBADF;
+        _cleanup_(pty_forward_freep) PTYForward *forward = NULL;
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        _cleanup_(sd_event_source_unrefp) sd_event_source *exit_source = NULL;
+        int r;
+
+        log_setup();
+
+        assert_se(sigprocmask_many(SIG_BLOCK, NULL, SIGCHLD) >= 0);
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        _cleanup_strv_free_ char **l = strv_copy(argv + optind);
+        if (!l)
+                return log_oom();
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
+
+        (void) sd_event_set_signal_exit(event, true);
+
+        pty_fd = openpt_allocate(O_RDWR|O_NOCTTY|O_NONBLOCK|O_CLOEXEC, /*ret_slave=*/ NULL);
+        if (pty_fd < 0)
+                return log_error_errno(pty_fd, "Failed to acquire pseudo tty: %m");
+
+        peer_fd = pty_open_peer(pty_fd, O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (peer_fd < 0)
+                return log_error_errno(peer_fd, "Failed to open pty peer: %m");
+
+        if (!arg_quiet)
+                log_info("Press ^] three times within 1s to disconnect TTY.");
+
+        r = pty_forward_new(event, pty_fd, /*flags=*/ 0, &forward);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create PTY forwarder: %m");
+
+        if (!isempty(arg_background))
+                (void) pty_forward_set_background_color(forward, arg_background);
+
+        pty_forward_set_handler(forward, pty_forward_handler, NULL);
+
+        r = pidref_safe_fork_full(
+                        "(sd-ptyfwd)",
+                        (int[]) { peer_fd, peer_fd, peer_fd },
+                        /*except_fds=*/ NULL,
+                        /*n_except_fds=*/ 0,
+                        /*flags=*/ FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_REARRANGE_STDIO,
+                        &pidref);
+        if (r < 0)
+                return log_error_errno(r, "Failed to fork child: %m");
+
+        if (r == 0) {
+                (void) execvp(l[0], l);
```

Comment:
you need setsid() here i guess so that you get a new posix session assigned. and then you need to make your new tty the controlling tty of that session (by calling TIOCSTTY ioctl on one of the stdin/stdout/stderr fds
)
the credentials(7) man page explains this briefly under "Process group ID and session ID".

Reply:
actually i think the setsid() + TIOCSCTTY thing should be done via a new ForkFlags flags, maybe just add FORK_SETSID and FORK_TIOCSCTTY, which done one of each flags.

maybe assert() if TIOCSTTY is used without FORK_SETSID, since this must fail, since only processes that are session leaders can call that ioctl. And we know for sure that if we fork something fresh off it's not going to be a session leader unless we already called setsid()